src/Build/BackEnd/Components/Caching/ResultsCache.cs

```diff
@@ -332,6 +333,15 @@ private static bool CheckResults(BuildResult result, List<string> targets, HashS
             return returnValue;
         }
 
+        /// <summary>
+        /// Checks results for the specified build flags.
+        /// </summary>
+        /// <param name="buildRequestDataFlags">The current request build flags.</param>
+        /// <param name="buildResultDataFlags">The existing build result data flags.</param>
+        /// <returns>False if there is any difference in the data flags that can cause missed build data, true otherwise.</returns>
+        private static bool CheckBuildDataFlagsResults(BuildRequestDataFlags buildRequestDataFlags, BuildRequestDataFlags buildResultDataFlags) =>
+            buildRequestDataFlags != BuildRequestDataFlags.ProvideProjectStateAfterBuild || buildResultDataFlags == buildRequestDataFlags;
```

Comment:
`BuildRequestDataFlags` are flags so this should look something like

```suggestion
            buildRequestDataFlags.HasFlag(BuildRequestDataFlags.ProvideProjectStateAfterBuild) == buildResultDataFlags.HasFLag(BuildRequestDataFlags.ProvideProjectStateAfterBuild);
```

but we will probably want to test for more flags, basically everything that can affect the results.

Reply:
@ladipro I believe the logic is that every flag other than `ProvideProjectStateAfterBuild` is considering as possible to affect results. 
However, by looking at existing flags, please consider it we shall test for `ProvideSubsetOfStateAfterBuild` as well. I mean only when both flags are not set we can consider this request cache satisfied as no results.
I am not sure if I understand it correctly, but to me check can be 
```C#
(!buildRequestDataFlags.HasFlag(BuildRequestDataFlags.ProvideProjectStateAfterBuild)  && 
 !buildRequestDataFlags.HasFlag(BuildRequestDataFlags.ProvideSubsetOfStateAfterBuild)) 
||
(buildResultDataFlags & 
 ~BuildRequestDataFlags.ProvideProjectStateAfterBuild  & 
 ~BuildRequestDataFlags.ProvideSubsetOfStateAfterBuild)  == 
(buildRequestDataFlags & 
 ~BuildRequestDataFlags.ProvideProjectStateAfterBuild & 
 ~BuildRequestDataFlags.ProvideSubsetOfStateAfterBuild)
 ```
 So changes in `Provide*StateAfterBuild` are ignored as those do not affect result.
 Additionally if no project or subset of project is going to be reported we can satisfy it by cache, but I do not understand this particular logic, as other changes in build request could affect resulting build artifacts or overall build success/failure (for example `IgnoreMissingEmptyAndInvalidImports` flag).

Reply:
@rainersigwald , are you aware of any other flags that we have to respect before extracting cache results?

Reply:
> @ladipro I believe the logic is that every flag other than ProvideProjectStateAfterBuild is considering as possible to affect results.

`ClearCachesAfterBuild` looks like an example of a "meta flag" which does not affect the results.

`ProvideSubsetOfStateAfterBuild` will likely need a slightly more sophisticated handling. Note that we never check for this flag in product code. We instead use non-null `RequestedProjectState` as the indicator that `ProjectStateAfterBuild` should be set. https://github.com/dotnet/msbuild/blob/5d7f5723223b823c9ed62ad52cc04e4c17a5848b/src/Build/BackEnd/Components/RequestBuilder/TargetBuilder.cs#L190-L200

Reply:
> > @ladipro I believe the logic is that every flag other than ProvideProjectStateAfterBuild is considering as possible to affect results.
> 
> `ClearCachesAfterBuild` looks like an example of a "meta flag" which does not affect the results.
> 
> `ProvideSubsetOfStateAfterBuild` will likely need a slightly more sophisticated handling. Note that we never check for this flag in product code. We instead use non-null `RequestedProjectState` as the indicator that `ProjectStateAfterBuild` should be set.
> 
> https://github.com/dotnet/msbuild/blob/5d7f5723223b823c9ed62ad52cc04e4c17a5848b/src/Build/BackEnd/Components/RequestBuilder/TargetBuilder.cs#L190-L200

Ladi, I made an attempt to add this flag explicitly if RequestedProjectState is passed. 
Is it wrong?

Reply:
It looks reasonable. Requiring API users to set this flag would be a breaking change so we can at least make the data consistent internally.

However, I'm not sure if this is enough to handle `ProvideSubsetOfStateAfterBuild` properly. If one build request asks for a certain subset of state and then another one for a different subset, then they both have the flag set but the expected build results are not the same and the cached entry should not be used.

Reply:
done